### PR TITLE
Fixes + mat sys improvements

### PR DESCRIPTION
**Pull request by HENDRIX-ZT2**
@niftools/blender-nif-plugin-reviewer 

# Overview
[Overview of the content of the pull request]
Expand texture functionality node support

##  Detailed Description
[List of functional updates]
Add back in Texture Export
- export_bs_shader_property 
- export_texturing_property
- Fixes issue import Oblivion mesh

Created dedicated slots for each texture type
- Dramatically simplify the texture_slot lookup code
- Use slot lookup when exporting NiTexturingProperty type
- Take path from the ShaderNode.

Fix issue with Menu registration for ExtraData

## Fixes Known Issues
[Ordered list of issues fixed by this PR]
#330 Export Zoo Tycoon 2 Error
#309 Export crashes if a root node exists, but isn't in selected objects
#232 Scene : Make the versioning system functional

## Documentation
[Overview of updates to documentation]

## Testing
[Overview of testing required to ensure functionality is correctly implemented]

### Manual
[Order steps to manually verify updates are working correctly]

### Automated
[List of tests run, updated or added to avoid future regressions]

## Additional Information
[Anything else you deem relevant]

### CODE REVIEW

**Comment by neomonkeus** [comment]
Looks good

File: io_scene_nif/modules/nif_export/property/texture/__init__.py [hunk 1]
```diff
@@ -53,46 +53,45 @@ def __init__(self):
         self.texture_anim = TextureAnimation()
         self.b_mat = None
 
-        # todo [texture] refactor texture grabbing to use the dict below
         self.slots = {}
         self._reset_fields()
-        self.b_diffuse_slot = None
```
Comment (neomonkeus) [inline io_scene_nif/modules/nif_export/property/texture/__init__.py:55]: ❤️

File: io_scene_nif/modules/nif_export/property/texture/__init__.py [hunk 2]
```diff
@@ -53,46 +53,45 @@ def __init__(self):
...
+        # check if a node is plugged into the b_texture_node's vector input
+        links = b_texture_node.inputs[0].links
+        if not links:
+            return 0
+        uv_node = links[0].from_node
+        if isinstance(uv_node, bpy.types.ShaderNodeUVMap):
+            uv_name = uv_node.uv_map
+            try:
+                # ignore the "UV" prefix
+                return int(uv_name[2:])
```
Comment (neomonkeus) [inline io_scene_nif/modules/nif_export/property/texture/__init__.py:85]: Will this always be UV prefixed.
Is this consistent for both when a uv map is created through import of existing mesh and a user created one.

File: io_scene_nif/modules/nif_import/property/texture/__init__.py [hunk 3]
```diff
@@ -204,11 +204,11 @@ def create_texture_slot(self, b_mat, n_tex_desc):
 
     def link_diffuse_node(self, b_texture_node):
         self.diffuse_texture = b_texture_node
-        b_texture_node.label = "Diffuse"
+        b_texture_node.label = "Base"
```
Comment (neomonkeus) [inline io_scene_nif/modules/nif_import/property/texture/__init__.py:207]: We should probably extract these to constants. 
Probably need to make them accessible via the ui|?

File: io_scene_nif/modules/nif_export/property/texture/__init__.py (hunk 2 above)
Comment (HENDRIX-ZT2) [reply io_scene_nif/modules/nif_export/property/texture/__init__.py:85]: We should just enforce this or any naming convention, and make warn the users if their uv names don't stick to it. We could of course index into the list of mesh uv layer names, but that seems to be more complicated than needed. A lot of the mesh - material entanglement resulted from devs being too cautious.

File: io_scene_nif/modules/nif_import/property/texture/__init__.py (hunk 3 above)
Comment (HENDRIX-ZT2) [reply io_scene_nif/modules/nif_import/property/texture/__init__.py:207]: It's exposed as a text entry in the node editor, we could probably turn it into a dropdown but I'm not sure.

File: io_scene_nif/modules/nif_export/property/texture/__init__.py (hunk 2 above)
Comment (neomonkeus) [reply io_scene_nif/modules/nif_export/property/texture/__init__.py:85]: Yeah I can understand that. 
There is also the balance between expecting it to exist and not, if it is a user issue or optional.

File: io_scene_nif/modules/nif_import/property/texture/__init__.py (hunk 3 above)
Comment (neomonkeus) [reply io_scene_nif/modules/nif_import/property/texture/__init__.py:207]: Yeah something to consider from UX pov.

File: io_scene_nif/modules/nif_import/property/shader/bsshaderproperty.py [hunk 4]
```diff
@@ -102,85 +103,83 @@ def register(self, processor):
...
+        # if hasattr(bs_shader_property, 'texture_clamp_mode'):
+        #     self.import_clamp(self.b_mat, bs_shader_property)
+        #
+        # if hasattr(bs_shader_property, 'uv_offset'):
+        #     self.import_uv_offset(self.b_mat, bs_shader_property)
+        #
+        # if hasattr(bs_shader_property, 'uv_scale'):
+        #     self.import_uv_scale(self.b_mat, bs_shader_property)
 
         # Diffuse color
```
Comment (neomonkeus) [inline io_scene_nif/modules/nif_import/property/shader/bsshaderproperty.py:129]: I think we will have to redo this whole section as the original code is convoluted by itself but we should look to have specific type.

Comment (HENDRIX-ZT2) [reply io_scene_nif/modules/nif_import/property/shader/bsshaderproperty.py:129]: Right, eventually we want to support the same for nitextureprops and use the shared material api class to take care of all the settings.

File: io_scene_nif/modules/nif_export/property/texture/__init__.py [hunk 5]
```diff
@@ -58,18 +58,18 @@ def __init__(self):
...
-            "Decal 0": None,
-            "Decal 1": None,
-            "Decal2": None,
+                "Base": None,
+                "Dark": None,
+                "Detail": None,
+                "Gloss": None,
+                "Glow": None,
+                "Bump Map": None,
+                "Decal 0": None,
```
Comment (neomonkeus) [inline io_scene_nif/modules/nif_export/property/texture/__init__.py:67]: One thing we may need to consider down the line, is that the Decal/Detail slots are a numbered array. 
I gleaned this from @Greatness7 so we can get more info from him at some point.

File: io_scene_nif/modules/nif_import/property/texture/types/nitextureprop.py [hunk 6]
```diff
@@ -59,81 +59,43 @@ def __init__(self):
...
+                "Detail": None,
+                "Gloss": None,
+                "Glow": None,
+                "Bump Map": None,
+                "Decal 0": None,
+                "Decal 1": None,
+                "Decal 2": None,
+                # extra shader stuff?
+                "Specular": None,
+                "Normal": None,
```
Comment (neomonkeus) [inline io_scene_nif/modules/nif_import/property/texture/types/nitextureprop.py:75]: Normal would be required for oblivion, it maps to the _n.dds textures.

**Comment by ajthedinosaurman** [comment]
Hi Hendrix. I'm having some trouble with importing bfb models into Blender 2.49b, and every time I select, import bfb, it pops up an error saying, "python script error, check console". I have everything installed. I would appreciate your help on this. Thank you, God bless you.

**Pull request merged**